rfcs/2022-03-15-11851-ingest-opentelemtry-traces.md

```diff
@@ -0,0 +1,399 @@
+this also involve evaluating a VRL condition on traces. The key problem here is to exposes traces and spans field in a
+way that the user can still manipulate those easily.
+
+This however raises the case of the granularity of a single event ; for instance multiple traces can bundle into a
+single payload in both OpenTelemetry and Datadog wire format. Enabling clear processing without ambiguity advocate for a
+clear constraint that should be enforced by all future traces sources : **a single Vector event shall not hold data
+relative to more that one trace**.
+
+A completely different usecase is traces sampling, but it cover two major variations:
+
+- Simple sampling: either cap/pace the trace flow at a given rate or sample 1 trace per 10/100/1000/etc. traces, and
+  this is already available thanks to the `sample` and `throttle` transforms
+- Outliers isolation, this would mean keeping some traces based on some advanced criteria, like execution time above
+  p99, this would require comparison against histogram / sketches.
+
+Another valuable identified usecase is the ability to provide seamless conversion between any kind of Vector supported
+traces, this means that the Vector internal traces representation shall be flexible enough to acomodate conversion
+to/from any trace format in sources and sinks that work with traces. Given the traction from the Opentelemetry project,
+and the fact that it [comes with a variety of fields][otlp-trace-proto-def] to cover most usecases.
+
+## Cross cutting concerns
+
+N/A
+
+## Scope
+
+### In scope
+
+- `opentelemetry` source, with both http and grpc support, decoding traces only, but with provision for other datatypes and emitting traces on a named output `traces`
+- Support `opentelemetry` source to `datadog_traces` sink forwarding by dealing with:
+  - Traces normalization to a single format inside Vector
+  - Conversion to/from this format in all traces sources/sinks
+- APM stats computation logic, with an implementation for the `opentelemetry` sources, applicable for all traces
+  sources. It will come with a knob to turn it on/off.
+
+### Out of scope
+
+N/A
+
+## Pain
+
+- Avoid complex setup when ingesting traces, ultimately pointing every tracing lib directly to Vector should just work
+  out-of-the-box with minimal config.
+
+## Proposal
+
+### User Experience
+
+- User would point OpenTelemetry tracing lib directly to a local Vector deployement
+- Vector would be configured with a minimal config looking like:
+
+```yaml
+sources:
+  otlp:
+    type: opentelemetry
+    address: "[::]:8081"
+    mode: grpc
+    traces_stats: true # Traces stats would be compliant with APM stats
+
+sinks:
+  dd_trace:
+    type: datadog_traces
+    default_api_key: 12345678abcdef
+    inputs:
+     - otlp.traces # Would exclusively emit traces
+     - otlp.traces_stats # Would exclusively emit metrics
+```
```

Comment:
This is clever, and I understand why metrics are generated at the source. My only concern is the footgun around not connecting the `traces_stats` input to the `datadog_traces` sink. It's not obvious or required and I'd expect most users to not do this. Given that the Datadog APM experience is diminished without the stats, we should think about how we can make this automatic or required for the `datadog_traces` sink.

Reply:
Yeah, doing it at the sink level is another option. However, for `datadog_agent` -> `datadog_traces` we do need to propagate the APM stats from the source as the Datadog Agent may not forward all traces and so calculating them in Vector will be inaccurate.

It would be an option for traces coming from the `opentelemetry` source though.

I'm still partial to introducing a transform to generate these stats as a first step. This is flexible enough to cover a myriad of use-cases and still leaves the door open for us to streamline it.